src/main/src/main/java/org/geoserver/security/decorators/SecuredFeatureSource.java

```diff
@@ -104,12 +105,37 @@ protected SecuredFeatureSource(FeatureSource<T, F> delegate, WrapperPolicy polic
                                         + "by security (because they are required by the schema). "
                                         + "Either the security setup is broken or you have a security breach");
                     }
-                    return SecuredObjects.secure(fc, policy);
+                    result = SecuredObjects.secure(fc, policy);
                 }
             } else {
-                return SecuredObjects.secure(fc, policy);
+                result = SecuredObjects.secure(fc, policy);
             }
         }
+        AccessLimits limits = policy.getLimits();
+        if (limits instanceof VectorAccessLimits) {
+            VectorAccessLimits vectorLimits = (VectorAccessLimits) limits;
+            result = decoratesForClipping(vectorLimits, result);
+        }
+        return result;
+    }
+
+    private FeatureCollection decoratesForClipping(
+            VectorAccessLimits limits, FeatureCollection collection) {
+        if (!(collection instanceof SimpleFeatureCollection))
+            throw new RuntimeException("Cannot clip on Complex Features");
+        Geometry clipFilter = limits.getClipVectorFilter();
+        Geometry intersectFilter = limits.getIntersectVectorFilter();
+        if (clipFilter != null) {
+            if (intersectFilter != null)
```

Comment:
As noted in the GeoTools PR, this approach is inefficient. Better to check if there is a intersectFilter while working in ``getReadQuery`` and "and" an intersection filter there, so that it send down to the data source for efficient processing.